Improve passthrough of -march and -mcpu from compiler to llvm-mca

peterwaller-arm opened the pull request:
This patch fixes a few issues with argument passthrough that have been
discovered since #6682 was landed.

* X86 didn't work if -march=CPU was supplied.
* It wasn't possible to override flags on MCA, because the auto-detected
  flags came before user flags.
* Since compiler-explorer runs on x86 native hardware, if targetting
  aarch64, it would result in an error about the default requested
  native CPU not being available for the architecture.

X86 clang uses -march to specify the CPU, and it's invalid to supply
-march=<cpu> to LLVM-MCA. Meanwhile, LLVM-MCA requires an -mcpu
argument.

Therefore, to handle x86 it's necessary to rewrite -march to -mcpu so
long as the march flag is used to supply a CPU value (i.e. not an
architecture level beginning with `x86-64`).

Per https://maskray.me/blog/2022-08-28-march-mcpu-mtune, I understand
x86 is currently the only target to have this behaviour so it needs
special casing.

Additionally, llvm-mca assumes -mcpu=native if it is not otherwise
specified. Since compiler-explorer runs on x86 hardware, it is easy to
end up in a situation for example targetting aarch64 but with -mcpu=<an
x86 CPU>. To handle this case some logic is added to detect if a CPU is
not specified and the target is not x86, in which case -mcpu=generic is
supplied instead as a reasonable default for the target.

Further, handle the case where --target is being passed to the compiler,
and pass this through as -mtriple.

Testing:

* I have been testing on an aarch64-native machine.
* I have tested targeting x86 by specifying --target and using the
  override functionality in compiler explorrer to specify x86.
* I have been inspecting the `--debug` output to see what flags are
  passed to LLVM-MCA.
* I've checked that I can override the -mcpu in MCA regardless of what
  is passed to the compiler.
* I have checked that I can supply either -march=x86-64* or -march=<cpu>
  on the x86 target and things behave as expected.

Fixes: #6709

vient commented on lib/tooling/llvm-mca-tool.ts:
```diff
@@ -47,21 +47,56 @@ export class LLVMMcaTool extends BaseTool {
+
+        for (const arg of compilationInfo.options) {
+            if (arg.startsWith('-march')) {
+                if (target?.startsWith('x86_64')) {
```
Isn't 32-bit case missing?
```
$ clang -###
Ubuntu clang version 18.1.3 (1ubuntu1)
Target: x86_64-pc-linux-gnu

$ clang -march=pentium -m32 -###
Ubuntu clang version 18.1.3 (1ubuntu1)
Target: i386-pc-linux-gnu
```

peterwaller-arm replied:
@vient Hm, I'm not confident about how to infer this without invoking clang and parsing the target which I'd prefer not to do (compiler explorer folks, is it OK to do that here?); is it sufficient to look for -m32 in the args or check if it is an i386 compiler (via `target`)?

vient replied:
It is sufficient with x86, don't know about other architectures.

peterwaller-arm replied:
Thanks. I've implemented it as discussed.

vient commented on lib/tooling/llvm-mca-tool.ts:
```diff
@@ -47,21 +47,56 @@ export class LLVMMcaTool extends BaseTool {
+        // compiler-explorer is running llvm-mca on, and it results in
+        // an error that MCA can't find the CPU for the target.
+        if (target && !target.startsWith('x86_64') && !haveCPU) {
+            prependArgs.push('-mcpu=generic');
```
I think `generic` is also a reasonable default for x86_64 as well (and x86 for the matter, so always). Otherwise llvm-mca successfully finds native target, and then results depend on a server which processed this run.

peterwaller-arm replied:
> A bit strange that compiler defaults to some generic target but llvm-mca defaults to native.

I take it that the compiler defaults to a generic CPU because this ensures wide compatibility of the compiled binaries (rather than CPU-specific), whereas MCA defaults to giving you a performance analysis of the machine the user is on, as a convenience. This makes reasonable sense to me but doesn't work if you're running MCA in a different environment than you are targetting, such as in on compiler-explorer.

I like the idea of using -mcpu=generic if mcpu (or on x86, march) is not otherwise specified on compiler-explorer.

peterwaller-arm replied:
Done.

peterwaller-arm commented:
Updated, please take another look.

vient commented on lib/tooling/llvm-mca-tool.ts:
```diff
@@ -47,21 +47,64 @@ export class LLVMMcaTool extends BaseTool {
+        let target = isets.getInstructionSetTarget(compilationInfo.compiler.instructionSet);
+        const prependArgs: string[] = [];
 
+        // Check if compiler target is overridden with --target.
```
Clang accepts arguments in two forms, `--target=x86_64-pc-linux-gnu` and `-target x86_64-pc-linux-gnu`

This quickly becomes more and more complicated 😅

peterwaller-arm replied:
Addressed.

vient commented on lib/tooling/llvm-mca-tool.ts:
```diff
@@ -47,21 +47,64 @@ export class LLVMMcaTool extends BaseTool {
+        let haveCPU = false;
+        let haveM32 = false;
+        for (const arg of compilationInfo.options) {
+            if (arg.startsWith('-march') && target?.startsWith('x86_64')) {
```
Target can start with `i386` which means 32-bit mode. You don't even need to pass `-m32` in this case, just passing `-target i386-pc-linux-gnu` makes clang compile 32-bit code.

Which means, you need to check for `x86_64` or `i386` here.

peterwaller-arm replied:
I put something in before seeing your edit. Do you think what I've done is sufficient or do you want to see more, and can you suggest what the conditions should be, and whether it is necessary? I'm still thinking in terms of avoiding invoking the compiler.

vient commented on lib/tooling/llvm-mca-tool.ts:
```diff
@@ -47,21 +47,64 @@ export class LLVMMcaTool extends BaseTool {
+            prependArgs.push('-mcpu=generic');
+        }
+
+        if (target?.startsWith('x86_64') && haveM32) {
```
So this handles one case where 32-bit mode is enabled via `-m32`. Need to check for another case when target triple starts with `i386`.

vient replied:
Right, we don't need to handle `i386` triple case - we have a triple after all, passing it is enough.

vient replied:
Btw, I'm using one-liner like this to test parameters `echo 'int main(int argc) { return argc * argc; }' | clang -x c - -target i386-pc-linux-gnu -S -o - | llvm-mca -mtriple=i386-pc-linux-gnu`

peterwaller-arm commented on lib/tooling/llvm-mca-tool.ts:
```diff
@@ -47,21 +47,64 @@ export class LLVMMcaTool extends BaseTool {
+                if (arg.startsWith('-march=x86-64')) {
+                    // This is an architecture level, not a CPU, so don't translate.
+                    continue;
+                }
```
Removed.

peterwaller-arm commented:
Thanks for the speedy review and attention to detail. I hope I've mostly addressed your concerns - I'll be back from Monday onwards if there are further things to address.

vient commented:
Looks good enough, I think? It's kinda hard to reason about this code, would be easier in the form of tests:
* From clang x86_64 `["-target", "i386"]` we expect to get llvm-mca parameters `["-march=x86", "-mcpu=generic"]`
* x86_64 `["-march=skylake", "-m32"]` -> `["-mcpu=skylake", "-march=x86"]`
* aarch64 ...

and so on

mattgodbolt commented on lib/tooling/llvm-mca-tool.ts:
```diff
@@ -47,21 +47,59 @@ export class LLVMMcaTool extends BaseTool {
+            if (arg === '-target') argIsTarget = true;
+        }
+
+        if (target != null) prependArgs.push('-mtriple=' + target);
```
should this be `!==` here? or are you trying to do something cunning in typescript/javascript that I don't understand? Is this materially different to `if (target) ...` ?

peterwaller-arm replied:
The choice of expression was arbitrary. I've written it as `if (target)` as I understand you are suggesting.

mattgodbolt reviewed (commented):
sorry for the slow review, have been on holiday. One question but otherwise LGTM... @ me and I'll merge/approve after you've had a chance to review

mattgodbolt reviewed (approved):
Thanks!
